src/Microsoft.Azure.Relay/RelayedHttpListenerResponse.cs

```diff
@@ -149,11 +154,51 @@ void Dispose(bool disposing)
             }
         }
 
-        void CheckDisposed()
+        internal void SetReadOnly()
+        {
+            this.readOnly = true;
+        }
+
+        void CheckDisposedOrReadOnly()
         {
             if (this.disposed)
             {
-                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(this.GetType().FullName), this);
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(this.ToString());
+                throw RelayEventSource.Log.ThrowingException(new ObjectDisposedException(message), this);
+            }
+
+            if (this.readOnly)
+            {
+                string message = this.Context.TrackingContext.EnsureTrackableMessage(SR.ResponseBodyStarted);
+                throw RelayEventSource.Log.ThrowingException(new InvalidOperationException(message), this);
+            }
+        }
+
+        class ResponseWebHeaderCollection : WebHeaderCollection
```

Comment:
>ResponseWebHeaderCollection [](start = 14, length = 27)

Looking at the API doc, I wonder if we need to also check for AddWithoutValidate() and Clear()

Reply:
Clear can be added. AddWithoutValidate seems to be a protected non-virtual method.

---
In reply to: [276867326](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276867326) [](ancestors = 276867326)

Reply:
Clear added in iteration 2.

---
In reply to: [276867837](https://github.com/Azure/azure-relay-dotnet/pull/105#discussion_r276867837) [](ancestors = 276867837,276867326)